### ISSUE TITLE
fix(ChangeDetection): ngAfterViewChecked is called when ChangeDetector is detached

### ISSUE THREAD

**Issue by PascalPrecht** [angular/angular#7054]
When detaching a change detector from the change detector tree, it should not cause the component to call `ngAfterViewChecked` or `ngDoCheck`. However, it seems those are called, while change detection itself is **not** performed.

Here's a plunk that illustrates the issue: http://plnkr.co/edit/px2oAkjzElRy3iqgP6jr?p=preview

**I'll explain really quick what it does:**

It's a single component that is bootstrapped and it detaches the change detector from the change detector tree:

```
  ngOnInit() {
    this.cd.detach();
  }
```

Whenever `ngAfterViewChecked` is called, I want to add a class to the component's host element so I can visualize that change detection has been performed. The class is then removed after 2 seconds. In order to make sure that `setTimeout()` doesn't trigger infinite `tick()` calls, this happens outside Angular's zone:

```
  ngAfterViewChecked() {
    this.el.nativeElement.classList.add('checked');
    this.zone.runOutsideAngular(() => {
      setTimeout(() => {
        this.el.nativeElement.classList.remove('checked');
      }, 2000);
    })
  }
```

**Expectation**: When bootstrapping the app, or clicking "Trigger CD", the app component shouldn't turn green. But it does. Although, we can see that CD itself is not performed, as the binding of the component is not reflected in the DOM

**UPDATE**:

The exact same happens to `ngDoCheck()` callback. Even though CD is detached, `ngDoCheck()` is called. Here's a punk that demonstrates it (pretty much the same thing, just that it turns red): http://plnkr.co/edit/HNRYZVdly2QG4MLHm64a?p=preview

**Comment by alfonso-presa** [comment]
I think that both of the issues you have opened the problem is that the `ngAfterViewChecked` is called from the parent of the ChangeDetector you're controlling with the OnPush strategy or with the detachment of the CD reference.

Instead of doing `this.cd.detach` do `this.cd._cd.parent.mode=ChangeDetectionStrategy.Detached` and you'll see that the behavior is the one you're expecting (although that code is obviously not correct).... 

I guess that the ChangeDetectorRef you get it's for the component's content and not for the component itself. This also applies to the other issue, as the OnPush strategy controls the component's content, and not the "component's container tag" ChangeDetector which is the one triggering the `ngAfterViewChecked` method.... Whether this is alright or not, I cannot tell :-).

Does it have sense?

**Comment by PascalPrecht** [comment]
> I think that in both of the issues you have opened the problem is that the ngAfterViewChecked is called from the parent of the ChangeDetector you're controlling with the OnPush strategy or with the detachment of the CD reference.

I'd assume that `ngAfterViewChecked()` is called when the **view** of that component has been checked by its ChangeDetector (because that's what it does, AFAIK).

> This also applies to the other issue, as the OnPush strategy controls the component's content, and not the "component's container tag"

By `content` you mean `view` and actually not `content?

> Whether this is alright or not, I cannot tell :-)

Well, that's why I opened this issue.

**Comment by alfonso-presa** [comment]
I was only trying to give my point of view to the issue and of course not trying to answer it myself...

I thought that actually getting notified a ViewCheck when a check was done in terms of:
* Parent CD: hey component CD, has something change?
* Component CD: Nope, every immutable input is equal to my prior check

was a feature, considering that trivial check a ViewCheck on itself. That is why I thought the `ngAfterViewChecked` was actually called from the "container" CD and not by the view CD (yes, sorry, I totally meant to say view).

I see it useful as it is right now in case you want to detach the change detector, and when notified a ViewCheck, delay the CD check to try to reduce the amount the refreshes in some scenarios. In case the ChangeDetectorRef you get injected is the same than the one invoking the `ngAfterViewChecked` then this kind of behavior would be impossible (at least as I think it should be implemented, may be there's other better way to do it... Should I be using `ngAfterContentChecked` for that?).

Either way, I'll shut up, and let the  people that actually know how this should work talk :-). Sorry for the disturb.

**Comment by PascalPrecht** [comment]
@alfonso-presa you're not disturbing, I'm happy to hear your thoughts. 

To be honest, I in fact have the feeling that I misunderstand `onPush` in the sense that it *does* check itself but only not its subtree (including expression in its view).

**Comment by alfonso-presa** [comment]
@PascalPrecht that will make what I call `"container" CD` to be the actual `view CD`, and what I call `view CD` to be the subtree.... Which makes muuuuch more sense ^_^.

**Comment by PascalPrecht** [comment]
I think I get your wording now. But then what would `ngAfterContentChecked()` be for? And if that's what's called by the parent, how come those are called for the root component too?

This seems very confusing to me. I think an official statement if `OnPush` only skips subtree **not** itself, would help a lot.

**Comment by alfonso-presa** [comment]
+1

**Comment by PascalPrecht** [comment]
@vsavkin confirmed that this is a bug. `ngAfterViewChecked()` shouldn't be called for a component that is either detached or set to `OnPush()`

**Comment by alfonso-presa** [comment]
Ok, Thanks!! I guess I should be using `ngAfterContentChecked` then :-).

**Comment by PascalPrecht** [comment]
No, I think non of them should be called when the component is set to `OnPush` and no input property has changed. Because it doesn't update neither, view or content.

**Comment by alfonso-presa** [comment]
This is an example of what I was trying to achieve by using `ngAfterXXXChecked` along with detached mode:
http://plnkr.co/edit/gcqcDY?p=preview

I'll try to solve it somehow else and if not possible file another issue as a requirement as it's clear enough how `ngAfterXXXChecked` should behave :-).

Thanks!!

**Comment by DrewynV** [comment]
__________________________________

Hi! We just wanted to let you know we receievd your email. Thank you for submitting your request. Our aim is to get back to you within 24 hours at the latest. If you have any additional information to add to your request, please reply to this email.


Also please visit our knowledgebase (http://www.support.magicdust.com.au) to browse for further info, we are adding to it all the time.

Or here is our industry report about the 11 Essentials for a Successful Website (http://www.magicdust.com.au/web-design/.

We look forward to helping you succeed online.

Thanks

The Team @ Magicdust Digital

http://support.magicdust.com.au/

Ph (02) 9979 7474
M-F | 9 am-5:00 pm AEST (AEDT in Summer) excluding Public Holidays

------------------------------------------------------------------------------------------------------------
This message was sent to  reply@reply.github.com in reference to Case #166466.
------------------------------------------------------------------------------------------------------------


[[54ad95e4cf252e131d9022d888c8c2085cc8fa3b-673073422]]

**Comment by DrewynV** [comment]
__________________________________

Hi! We just wanted to let you know we receievd your email. Thank you for submitting your request. Our aim is to get back to you within 24 hours at the latest. If you have any additional information to add to your request, please reply to this email.


Also please visit our knowledgebase (http://www.support.magicdust.com.au) to browse for further info, we are adding to it all the time.

Or here is our industry report about the 11 Essentials for a Successful Website (http://www.magicdust.com.au/web-design/.

We look forward to helping you succeed online.

Thanks

The Team @ Magicdust Digital

http://support.magicdust.com.au/

Ph (02) 9979 7474
M-F | 9 am-5:00 pm AEST (AEDT in Summer) excluding Public Holidays

------------------------------------------------------------------------------------------------------------
This message was sent to  reply@reply.github.com in reference to Case #166464.
------------------------------------------------------------------------------------------------------------


[[b405cd2488070f20af0e629ab4a17adee65a58b0-673073420]]

**Comment by hadar111** [comment]
Anyone pass that out ? or have any walk around in order to avoid those checks ?

**Comment by tbosch** [comment]
So `OnPush` / `changeDetectorRef.detach()` only affect the template of a component. And this makes sense: 

Not calling the lifecycle hooks on a detached component would mean we probably also should not call set properties on the component.

You want to tell Angular that it does not need to dirty check the template as the components knows that e.g. all data is immutable. However, the components needs to be told when it gets new data so that it can tell Angular to dirty check the template again. 

For `ngAfterViewChecked` its debatable whether to call it or not for a detached change detector. However, special casing it also seems wrong.

I am closing this issue as it seems to work as intended.
